File: veriblock-core/src/test/java/org/veriblock/core/utilities/BitStringReaderTests.java
// VeriBlock NodeCore
// Copyright 2017-2020 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.

package org.veriblock.core.utilities;

import org.junit.Assert;
import org.junit.Test;
import org.veriblock.core.types.BitString;

public class BitStringReaderTests {
    @Test
    public void bitStringReaderTestSingleByte1() {
        // 01000000
        byte[] bytes = new byte[]{(byte)0x40};

        BitString bitString = new BitString(bytes);

        BitStringReader reader = new BitStringReader(bitString);
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(1), new byte[]{(byte)0x00}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(7), new byte[]{(byte)0x40}));

        reader = new BitStringReader(bitString);
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(2), new byte[]{(byte)0x01}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(6), new byte[]{(byte)0x00}));

        reader = new BitStringReader(bitString);
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(3), new byte[]{(byte)0x02}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(5), new byte[]{(byte)0x00}));

        reader = new BitStringReader(bitString);
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(4), new byte[]{(byte)0x04}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(4), new byte[]{(byte)0x00}));

        reader = new BitStringReader(bitString);
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(5), new byte[]{(byte)0x08}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(3), new byte[]{(byte)0x00}));

        reader = new BitStringReader(bitString);
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(6), new byte[]{(byte)0x10}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(2), new byte[]{(byte)0x00}));

        reader = new BitStringReader(bitString);
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(7), new byte[]{(byte)0x20}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(1), new byte[]{(byte)0x00}));

        reader = new BitStringReader(bitString);
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(8), new byte[]{(byte)0x40}));

        reader = new BitStringReader(bitString);
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(3), new byte[]{(byte)0x02}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(2), new byte[]{(byte)0x00}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(3), new byte[]{(byte)0x00}));

        try {
            reader = new BitStringReader(bitString);
            reader.readBits(9);
            Assert.fail();
        } catch (Exception e) { }

        reader = new BitStringReader(bitString);
        Assert.assertEquals(reader.remaining(), 8);
        Assert.assertEquals(reader.readBit(), 0x00);
        Assert.assertEquals(reader.remaining(), 7);
        Assert.assertEquals(reader.readBit(), 0x01);
        Assert.assertEquals(reader.remaining(), 6);
        Assert.assertEquals(reader.readBit(), 0x00);
        Assert.assertEquals(reader.remaining(), 5);
        Assert.assertEquals(reader.readBit(), 0x00);
        Assert.assertEquals(reader.remaining(), 4);
        Assert.assertEquals(reader.readBit(), 0x00);
        Assert.assertEquals(reader.remaining(), 3);
        Assert.assertEquals(reader.readBit(), 0x00);
        Assert.assertEquals(reader.remaining(), 2);
        Assert.assertEquals(reader.readBit(), 0x00);
        Assert.assertEquals(reader.remaining(), 1);
        Assert.assertEquals(reader.readBit(), 0x00);
        Assert.assertEquals(reader.remaining(), 0);
    }

    @Test
    public void bitStringReaderTestSingleByte2() {
        // 10001010
        byte[] bytes = new byte[]{(byte)0x8A};

        BitString bitString = new BitString(bytes);

        BitStringReader reader = new BitStringReader(bitString);
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(1), new byte[]{(byte)0x01}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(7), new byte[]{(byte)0x0A}));

        reader = new BitStringReader(bitString);
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(2), new byte[]{(byte)0x02}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(6), new byte[]{(byte)0x0A}));

        reader = new BitStringReader(bitString);
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(3), new byte[]{(byte)0x04}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(5), new byte[]{(byte)0x0A}));

        reader = new BitStringReader(bitString);
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(4), new byte[]{(byte)0x08}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(4), new byte[]{(byte)0x0A}));

        reader = new BitStringReader(bitString);
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(5), new byte[]{(byte)0x11}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(3), new byte[]{(byte)0x02}));

        reader = new BitStringReader(bitString);
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(6), new byte[]{(byte)0x22}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(2), new byte[]{(byte)0x02}));

        reader = new BitStringReader(bitString);
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(7), new byte[]{(byte)0x45}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(1), new byte[]{(byte)0x00}));

        reader = new BitStringReader(bitString);
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(8), new byte[]{(byte)0x8A}));

        reader = new BitStringReader(bitString);
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(1), new byte[]{(byte)0x01}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(4), new byte[]{(byte)0x01}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(2), new byte[]{(byte)0x01}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(1), new byte[]{(byte)0x00}));

        try {
            reader = new BitStringReader(bitString);
            reader.readBits(9);
            Assert.fail();
        } catch (Exception e) { }

        reader = new BitStringReader(bitString);
        Assert.assertEquals(reader.remaining(), 8);
        Assert.assertEquals(reader.readBit(), 0x01);
        Assert.assertEquals(reader.remaining(), 7);
        Assert.assertEquals(reader.readBit(), 0x00);
        Assert.assertEquals(reader.remaining(), 6);
        Assert.assertEquals(reader.readBit(), 0x00);
        Assert.assertEquals(reader.remaining(), 5);
        Assert.assertEquals(reader.readBit(), 0x00);
        Assert.assertEquals(reader.remaining(), 4);
        Assert.assertEquals(reader.readBit(), 0x01);
        Assert.assertEquals(reader.remaining(), 3);
        Assert.assertEquals(reader.readBit(), 0x00);
        Assert.assertEquals(reader.remaining(), 2);
        Assert.assertEquals(reader.readBit(), 0x01);
        Assert.assertEquals(reader.remaining(), 1);
        Assert.assertEquals(reader.readBit(), 0x00);
        Assert.assertEquals(reader.remaining(), 0);
    }

    @Test
    public void bitStringReaderTestDoubleByte1() {
        // 0011000101101101
        byte[] bytes = new byte[]{(byte)0x31, (byte)0x6D};

        BitString bitString = new BitString(bytes);

        BitStringReader reader = new BitStringReader(bitString);
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(1), new byte[]{(byte)0x00}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(15), new byte[]{(byte)0x31, (byte)0x6D}));

        reader = new BitStringReader(bitString);
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(2), new byte[]{(byte)0x00}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(14), new byte[]{(byte)0x31, (byte)0x6D}));

        reader = new BitStringReader(bitString);
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(3), new byte[]{(byte)0x01}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(13), new byte[]{(byte)0x11, (byte)0x6D}));

        reader = new BitStringReader(bitString);
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(4), new byte[]{(byte)0x03}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(12), new byte[]{(byte)0x01, (byte)0x6D}));

        reader = new BitStringReader(bitString);
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(5), new byte[]{(byte)0x06}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(11), new byte[]{(byte)0x01, (byte)0x6D}));

        reader = new BitStringReader(bitString);
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(6), new byte[]{(byte)0x0C}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(10), new byte[]{(byte)0x01, (byte)0x6D}));

        reader = new BitStringReader(bitString);
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(7), new byte[]{(byte)0x18}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(9), new byte[]{(byte)0x01, (byte)0x6D}));

        reader = new BitStringReader(bitString);
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(8), new byte[]{(byte)0x31}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(8), new byte[]{(byte)0x6D}));

        reader = new BitStringReader(bitString);
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(3), new byte[]{(byte)0x01}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(2), new byte[]{(byte)0x02}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(5), new byte[]{(byte)0x05}));
        Assert.assertTrue(Utility.byteArraysAreEqual(reader.readBits(6), new byte[]{(byte)0x2D}));

        try {
            reader = new BitStringReader(bitString);
            reader.readBits(5);
            reader.readBits(10);
            reader.readBits(2);
            Assert.fail();
        } catch (Exception e) { }

        reader = new BitStringReader(bitString);
        Assert.assertEquals(reader.remaining(), 16);
        Assert.assertEquals(reader.readBit(), 0x00);
        Assert.assertEquals(reader.remaining(), 15);
        Assert.assertEquals(reader.readBit(), 0x00);
        Assert.assertEquals(reader.remaining(), 14);
        Assert.assertEquals(reader.readBit(), 0x01);
        Assert.assertEquals(reader.remaining(), 13);
        Assert.assertEquals(reader.readBit(), 0x01);
        Assert.assertEquals(reader.remaining(), 12);
        Assert.assertEquals(reader.readBit(), 0x00);
        Assert.assertEquals(reader.remaining(), 11);
        Assert.assertEquals(reader.readBit(), 0x00);
        Assert.assertEquals(reader.remaining(), 10);
        Assert.assertEquals(reader.readBit(), 0x00);
        Assert.assertEquals(reader.remaining(), 9);
        Assert.assertEquals(reader.readBit(), 0x01);
        Assert.assertEquals(reader.remaining(), 8);
        Assert.assertEquals(reader.readBit(), 0x00);
        Assert.assertEquals(reader.remaining(), 7);
        Assert.assertEquals(reader.readBit(), 0x01);
        Assert.assertEquals(reader.remaining(), 6);
        Assert.assertEquals(reader.readBit(), 0x01);
        Assert.assertEquals(reader.remaining(), 5);
        Assert.assertEquals(reader.readBit(), 0x00);
        Assert.assertEquals(reader.remaining(), 4);
        Assert.assertEquals(reader.readBit(), 0x01);
        Assert.assertEquals(reader.remaining(), 3);
        Assert.assertEquals(reader.readBit(), 0x01);
        Assert.assertEquals(reader.remaining(), 2);
        Assert.assertEquals(reader.readBit(), 0x00);
        Assert.assertEquals(reader.remaining(), 1);
        Assert.assertEquals(reader.readBit(), 0x01);
        Assert.assertEquals(reader.remaining(), 0);
    }
}

File: veriblock-core/src/test/java/org/veriblock/core/utilities/BitStringReaderTests.kt
// VeriBlock NodeCore
// Copyright 2017-2021 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.
package org.veriblock.core.utilities

import io.kotest.assertions.fail
import io.kotest.matchers.shouldBe
import org.junit.Test
import org.veriblock.core.types.BitString

class BitStringReaderTests {
    @Test
    fun bitStringReaderTestSingleByte1() {
        // 01000000
        val bytes = byteArrayOf(0x40.toByte())
        val bitString = BitString(bytes)
        var reader = BitStringReader(bitString)
        Utility.byteArraysAreEqual(reader.readBits(1), byteArrayOf(0x00.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(7), byteArrayOf(0x40.toByte())) shouldBe true
        reader = BitStringReader(bitString)
        Utility.byteArraysAreEqual(reader.readBits(2), byteArrayOf(0x01.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(6), byteArrayOf(0x00.toByte())) shouldBe true
        reader = BitStringReader(bitString)
        Utility.byteArraysAreEqual(reader.readBits(3), byteArrayOf(0x02.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(5), byteArrayOf(0x00.toByte())) shouldBe true
        reader = BitStringReader(bitString)
        Utility.byteArraysAreEqual(reader.readBits(4), byteArrayOf(0x04.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(4), byteArrayOf(0x00.toByte())) shouldBe true
        reader = BitStringReader(bitString)
        Utility.byteArraysAreEqual(reader.readBits(5), byteArrayOf(0x08.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(3), byteArrayOf(0x00.toByte())) shouldBe true
        reader = BitStringReader(bitString)
        Utility.byteArraysAreEqual(reader.readBits(6), byteArrayOf(0x10.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(2), byteArrayOf(0x00.toByte())) shouldBe true
        reader = BitStringReader(bitString)
        Utility.byteArraysAreEqual(reader.readBits(7), byteArrayOf(0x20.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(1), byteArrayOf(0x00.toByte())) shouldBe true
        reader = BitStringReader(bitString)
        Utility.byteArraysAreEqual(reader.readBits(8), byteArrayOf(0x40.toByte())) shouldBe true
        reader = BitStringReader(bitString)
        Utility.byteArraysAreEqual(reader.readBits(3), byteArrayOf(0x02.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(2), byteArrayOf(0x00.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(3), byteArrayOf(0x00.toByte())) shouldBe true
        try {
            reader = BitStringReader(bitString)
            reader.readBits(9)
            fail("")
        } catch (e: Exception) {
        }
        reader = BitStringReader(bitString)
        reader.remaining().toLong() shouldBe 8
        reader.readBit().toLong() shouldBe 0x00
        reader.remaining().toLong() shouldBe 7
        reader.readBit().toLong() shouldBe 0x01
        reader.remaining().toLong() shouldBe 6
        reader.readBit().toLong() shouldBe 0x00
        reader.remaining().toLong() shouldBe 5
        reader.readBit().toLong() shouldBe 0x00
        reader.remaining().toLong() shouldBe 4
        reader.readBit().toLong() shouldBe 0x00
        reader.remaining().toLong() shouldBe 3
        reader.readBit().toLong() shouldBe 0x00
        reader.remaining().toLong() shouldBe 2
        reader.readBit().toLong() shouldBe 0x00
        reader.remaining().toLong() shouldBe 1
        reader.readBit().toLong() shouldBe 0x00
        reader.remaining().toLong() shouldBe 0
    }

    @Test
    fun bitStringReaderTestSingleByte2() {
        // 10001010
        val bytes = byteArrayOf(0x8A.toByte())
        val bitString = BitString(bytes)
        var reader = BitStringReader(bitString)
        Utility.byteArraysAreEqual(reader.readBits(1), byteArrayOf(0x01.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(7), byteArrayOf(0x0A.toByte())) shouldBe true
        reader = BitStringReader(bitString)
        Utility.byteArraysAreEqual(reader.readBits(2), byteArrayOf(0x02.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(6), byteArrayOf(0x0A.toByte())) shouldBe true
        reader = BitStringReader(bitString)
        Utility.byteArraysAreEqual(reader.readBits(3), byteArrayOf(0x04.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(5), byteArrayOf(0x0A.toByte())) shouldBe true
        reader = BitStringReader(bitString)
        Utility.byteArraysAreEqual(reader.readBits(4), byteArrayOf(0x08.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(4), byteArrayOf(0x0A.toByte())) shouldBe true
        reader = BitStringReader(bitString)
        Utility.byteArraysAreEqual(reader.readBits(5), byteArrayOf(0x11.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(3), byteArrayOf(0x02.toByte())) shouldBe true
        reader = BitStringReader(bitString)
        Utility.byteArraysAreEqual(reader.readBits(6), byteArrayOf(0x22.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(2), byteArrayOf(0x02.toByte())) shouldBe true
        reader = BitStringReader(bitString)
        Utility.byteArraysAreEqual(reader.readBits(7), byteArrayOf(0x45.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(1), byteArrayOf(0x00.toByte())) shouldBe true
        reader = BitStringReader(bitString)
        Utility.byteArraysAreEqual(reader.readBits(8), byteArrayOf(0x8A.toByte())) shouldBe true
        reader = BitStringReader(bitString)
        Utility.byteArraysAreEqual(reader.readBits(1), byteArrayOf(0x01.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(4), byteArrayOf(0x01.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(2), byteArrayOf(0x01.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(1), byteArrayOf(0x00.toByte())) shouldBe true
        try {
            reader = BitStringReader(bitString)
            reader.readBits(9)
            fail("")
        } catch (e: Exception) {
        }
        reader = BitStringReader(bitString)
        reader.remaining().toLong() shouldBe 8
        reader.readBit().toLong() shouldBe 0x01
        reader.remaining().toLong() shouldBe 7
        reader.readBit().toLong() shouldBe 0x00
        reader.remaining().toLong() shouldBe 6
        reader.readBit().toLong() shouldBe 0x00
        reader.remaining().toLong() shouldBe 5
        reader.readBit().toLong() shouldBe 0x00
        reader.remaining().toLong() shouldBe 4
        reader.readBit().toLong() shouldBe 0x01
        reader.remaining().toLong() shouldBe 3
        reader.readBit().toLong() shouldBe 0x00
        reader.remaining().toLong() shouldBe 2
        reader.readBit().toLong() shouldBe 0x01
        reader.remaining().toLong() shouldBe 1
        reader.readBit().toLong() shouldBe 0x00
        reader.remaining().toLong() shouldBe 0
    }

    @Test
    fun bitStringReaderTestDoubleByte1() {
        // 0011000101101101
        val bytes = byteArrayOf(0x31.toByte(), 0x6D.toByte())
        val bitString = BitString(bytes)
        var reader = BitStringReader(bitString)
        Utility.byteArraysAreEqual(reader.readBits(1), byteArrayOf(0x00.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(15), byteArrayOf(0x31.toByte(), 0x6D.toByte())) shouldBe true
        reader = BitStringReader(bitString)
        Utility.byteArraysAreEqual(reader.readBits(2), byteArrayOf(0x00.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(14), byteArrayOf(0x31.toByte(), 0x6D.toByte())) shouldBe true
        reader = BitStringReader(bitString)
        Utility.byteArraysAreEqual(reader.readBits(3), byteArrayOf(0x01.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(13), byteArrayOf(0x11.toByte(), 0x6D.toByte())) shouldBe true
        reader = BitStringReader(bitString)
        Utility.byteArraysAreEqual(reader.readBits(4), byteArrayOf(0x03.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(12), byteArrayOf(0x01.toByte(), 0x6D.toByte())) shouldBe true
        reader = BitStringReader(bitString)
        Utility.byteArraysAreEqual(reader.readBits(5), byteArrayOf(0x06.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(11), byteArrayOf(0x01.toByte(), 0x6D.toByte())) shouldBe true
        reader = BitStringReader(bitString)
        Utility.byteArraysAreEqual(reader.readBits(6), byteArrayOf(0x0C.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(10), byteArrayOf(0x01.toByte(), 0x6D.toByte())) shouldBe true
        reader = BitStringReader(bitString)
        Utility.byteArraysAreEqual(reader.readBits(7), byteArrayOf(0x18.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(9), byteArrayOf(0x01.toByte(), 0x6D.toByte())) shouldBe true
        reader = BitStringReader(bitString)
        Utility.byteArraysAreEqual(reader.readBits(8), byteArrayOf(0x31.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(8), byteArrayOf(0x6D.toByte())) shouldBe true
        reader = BitStringReader(bitString)
        Utility.byteArraysAreEqual(reader.readBits(3), byteArrayOf(0x01.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(2), byteArrayOf(0x02.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(5), byteArrayOf(0x05.toByte())) shouldBe true
        Utility.byteArraysAreEqual(reader.readBits(6), byteArrayOf(0x2D.toByte())) shouldBe true
        try {
            reader = BitStringReader(bitString)
            reader.readBits(5)
            reader.readBits(10)
            reader.readBits(2)
            fail("")
        } catch (e: Exception) {
        }
        reader = BitStringReader(bitString)
        reader.remaining().toLong() shouldBe 16
        reader.readBit().toLong() shouldBe 0x00
        reader.remaining().toLong() shouldBe 15
        reader.readBit().toLong() shouldBe 0x00
        reader.remaining().toLong() shouldBe 14
        reader.readBit().toLong() shouldBe 0x01
        reader.remaining().toLong() shouldBe 13
        reader.readBit().toLong() shouldBe 0x01
        reader.remaining().toLong() shouldBe 12
        reader.readBit().toLong() shouldBe 0x00
        reader.remaining().toLong() shouldBe 11
        reader.readBit().toLong() shouldBe 0x00
        reader.remaining().toLong() shouldBe 10
        reader.readBit().toLong() shouldBe 0x00
        reader.remaining().toLong() shouldBe 9
        reader.readBit().toLong() shouldBe 0x01
        reader.remaining().toLong() shouldBe 8
        reader.readBit().toLong() shouldBe 0x00
        reader.remaining().toLong() shouldBe 7
        reader.readBit().toLong() shouldBe 0x01
        reader.remaining().toLong() shouldBe 6
        reader.readBit().toLong() shouldBe 0x01
        reader.remaining().toLong() shouldBe 5
        reader.readBit().toLong() shouldBe 0x00
        reader.remaining().toLong() shouldBe 4
        reader.readBit().toLong() shouldBe 0x01
        reader.remaining().toLong() shouldBe 3
        reader.readBit().toLong() shouldBe 0x01
        reader.remaining().toLong() shouldBe 2
        reader.readBit().toLong() shouldBe 0x00
        reader.remaining().toLong() shouldBe 1
        reader.readBit().toLong() shouldBe 0x01
        reader.remaining().toLong() shouldBe 0
    }
}
